src/util.ts

```diff
@@ -828,3 +799,82 @@ export function parseMatrixInput(
   }
   return JSON.parse(matrixInput);
 }
+
+function removeDuplicateLocations(locations: SarifLocation[]): SarifLocation[] {
+  const newJsonLocations = new Set<string>();
+  return locations.filter((location) => {
+    const jsonLocation = JSON.stringify(location);
+    if (!newJsonLocations.has(jsonLocation)) {
+      newJsonLocations.add(jsonLocation);
+      return true;
+    }
+    return false;
+  });
+}
+
+export function fixInvalidNotifications(
+  sarif: SarifFile,
+  logger: Logger
+): SarifFile {
+  if (process.env[CODEQL_ACTION_DISABLE_DUPLICATE_LOCATION_FIX] === "true") {
+    logger.info(
+      "SARIF notification object duplicate location fix disabled by the " +
+        `${CODEQL_ACTION_DISABLE_DUPLICATE_LOCATION_FIX} environment variable.`
+    );
+    return sarif;
+  }
+  if (!(sarif.runs instanceof Array)) {
```

Comment:
Minor: using `Array.isArray` is generally better. However, in this particular case `instanceof` is fine. It's only really necessary when using iframes in DOM.

```suggestion
  if (!Array.isArray(sarif.runs)) {
```

https://stackoverflow.com/questions/22289727/difference-between-using-array-isarray-and-instanceof-array